fvm/transactionEnv.go

```diff
@@ -0,0 +1,993 @@
+package fvm
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/onflow/cadence"
+	jsoncdc "github.com/onflow/cadence/encoding/json"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/opentracing/opentracing-go"
+	traceLog "github.com/opentracing/opentracing-go/log"
+
+	"github.com/onflow/flow-go/fvm/blueprints"
+	"github.com/onflow/flow-go/fvm/crypto"
+	"github.com/onflow/flow-go/fvm/errors"
+	"github.com/onflow/flow-go/fvm/handler"
+	"github.com/onflow/flow-go/fvm/programs"
+	"github.com/onflow/flow-go/fvm/state"
+	"github.com/onflow/flow-go/fvm/utils"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
+	"github.com/onflow/flow-go/storage"
+)
+
+var _ runtime.Interface = &TransactionEnv{}
+
+// TransactionEnv is a read-write environment used for executing flow transactions.
+type TransactionEnv struct {
+	vm               *VirtualMachine
+	ctx              Context
+	sth              *state.StateHolder
+	programs         *handler.ProgramsHandler
+	accounts         *state.Accounts
+	uuidGenerator    *state.UUIDGenerator
+	contracts        *handler.ContractHandler
+	accountKeys      *handler.AccountKeyHandler
+	metrics          *handler.MetricsHandler
+	eventHandler     *handler.EventHandler
+	addressGenerator flow.AddressGenerator
+	rng              *rand.Rand
+	logs             []string
+	totalGasUsed     uint64
+	tx               *flow.TransactionBody
+	txIndex          uint32
+	txID             flow.Identifier
+	traceSpan        opentracing.Span
+	authorizers      []runtime.Address
+}
+
+func NewTransactionEnvironment(
+	ctx Context,
+	vm *VirtualMachine,
+	sth *state.StateHolder,
+	programs *programs.Programs,
+	tx *flow.TransactionBody,
+	txIndex uint32,
+	traceSpan opentracing.Span,
+) *TransactionEnv {
+
+	accounts := state.NewAccounts(sth)
+	generator := state.NewStateBoundAddressGenerator(sth, ctx.Chain)
+	uuidGenerator := state.NewUUIDGenerator(sth)
+
+	programsHandler := handler.NewProgramsHandler(
+		programs, sth,
+	)
+
+	// TODO set the flags on context
+	eventHandler := handler.NewEventHandler(ctx.Chain,
+		ctx.EventCollectionEnabled,
```

Comment:
Ideally, instead of controlling behaviour with switches it would be better to have different implementations here - if collection is not enabled, something like `NoopEventsCollector` etc.

Reply:
Exactly, that is what is in the list of future PRs, I didn't want to change the logic that much in this PR.